src/impression.js

```diff
@@ -53,74 +53,119 @@ export function resetTrackImpressionPromiseForTesting() {
 export function maybeTrackImpression(win) {
   let resolveImpression;
 
-  trackImpressionPromise = new Promise(resolve => {
+  const promise = new Promise(resolve => {
     resolveImpression = resolve;
   });
 
-  const viewer = Services.viewerForDoc(win.document);
-  viewer.isTrustedViewer().then(isTrusted => {
+  trackImpressionPromise = Services.timerFor(win).timeoutPromise(TIMEOUT_VALUE,
+      promise, 'TrackImpressionPromise timeout').catch(error => {
+        dev().warn('IMPRESSION', error);
+      });
+
+  Services.viewerForDoc(win.document).isTrustedViewer().then(isTrusted => {
     // Currently this feature is launched for trusted viewer, but still
     // experiment guarded for all AMP docs.
     if (!isTrusted && !isExperimentOn(win, 'alp')) {
       resolveImpression();
       return;
     }
 
-    trackImpressionFromClickParam(win, viewer, resolveImpression);
+    const replaceUrlPromise = handleReplaceUrl(win);
+    const clickUrlPromise = handleClickUrl(win);
+
+    Promise.all([replaceUrlPromise, clickUrlPromise]).then(() => {
+      resolveImpression();
+    }, () => {});
   });
 }
 
+/**
+ * Signal that impression tracking is not relevant in this environment.
+ */
+export function doNotTrackImpression() {
+  trackImpressionPromise = Promise.resolve();
+}
+
+/**
+ * Handle the getReplaceUrl and return a promise when url is replaced
+ * Only handles replaceUrl when viewer indicates AMP to do so. Viewer should indicate
+ * by setting the legacy replaceUrl init param and add `replaceUrl` to its capability param.
+ * Future plan is to change the type of legacy init replaceUrl param from url string
+ * to boolean value.
+ * Please NOTE replaceUrl and adLocation will never arrive at same time,
+ * so there is no race condition on the order of handling url replacement.
+ * @param {!Window} win
+ * @return {!Promise}
+ */
+function handleReplaceUrl(win) {
+  const viewer = Services.viewerForDoc(win.document);
+
+  // ReplaceUrl substitution doesn't have to wait until the document is visible
+  if (!viewer.getParam('replaceUrl')) {
```

Comment:
I don't understand why we still need to provide a `replaceUrl` param if we have the capability?

Reply:
Since we don't want to introduce the idea of per-document capability, we need a signal from viewer to indicate AMP runtime to ask for `replaceUrl`. The decision here is to use both the legacy `replaceUrl` init param and the capability to indicate that AMP runtime should and can ask viewer for async `replaceUrl`.

Reply:
Aren't the capabilities passed as a param (`cap`) to the document? I don't see how that's per-document any more than `replaceUrl`

Reply:
> Aren't the capabilities passed as a param (cap) to the document?

Yes, they are.

I think the question here is how viewer and AMP runtime agree on the communication protocol. `cap` can serve as a `per-document capability but we simply don't want the param to be changed across different docs for same viewer. That's why we also use the `replaceUrl` to decide to call `getReplaceUrl` or not. Does that make sense?

Reply:
> That's why we also use the replaceUrl to decide to call getReplaceUrl or not. Does that make sense?

No, since I can return anything from the request. You've only made an implicit dependency that we'll trip over later.

I'm think we should remove the `replaceUrl` check. If viewer says it has the capability, ask it for the url. If it doesn't, and it provides a `replaceUrl`, use that as a legacy fallback. Otherwise, there's no replacement.

Reply:
AMP has the capability to replaceUrl, it will ALWAYS state that it has the capability. However AMP should not always ask for a replaceUrl. As I stated before `replaceUrl` is a special param that placed by search Ads. If they don't place it at first, why we AMP runtime bother asking for something that's not necessary. 
As say init param `replaceUrl` indicates that this is a landing page for search ads. And AMP should try to replace url.
`cap` indicates the viewer's capability to provide async `replaceUrl`.

Reply:
> If they don't place it at first, why would AMP runtime bother asking for something that's not necessary.

That's my point entirely. You're made a dependency that's not obvious, and unnecessary. 

> However AMP should not always ask for a replaceUrl.  If they don't place it at first, why would AMP runtime bother asking for something that's not necessary.

It's not asked for because this is the only place that uses it. If we hit this code, we're in a search ad, right? It's a circle, one proves the other.

Reply:
> You're made a dependency that's not obvious, and unnecessary.

I agree the dependency is not very obvious here. I've tried the best to document in comments : (

> If we hit this code, we're in a search ad, right? It's a circle, one proves the other.
No that's not correct. The function `#maybeTrackImpression()` is called by AMP runtime regardless. https://github.com/ampproject/amphtml/blob/master/src/amp.js#L81

So I guess we are good here? That `replaceUrl` here serves as an indicator to we're in a landing page from search ad. And we should ask viewer for async `replaceUrl` if they support it.